File: src/clojure/examples/circle_area.clj
(ns examples.circle-area
  (:require [clj-see.program :as program])
  (:require [clojure.math.numeric-tower :refer [abs expt]]))

(defn circle-area [radius]
  (* radius radius Math/PI))

(defn sqr [x]
  (* x x))

; TODO: fitness can be cached, probably
(defn fitness [program]
  (let [program-size (count (program/all-paths program))
        diff-fn #(abs (- (circle-area %) (program %)))
        differences (map diff-fn (range 10))]
    (* program-size (- (apply + (map sqr differences))))))

(defmacro r [] `(first ~'args))

(def all-mutations
  {:non-terminal [(fn [_] (if (> (count _) 1) (rand-nth (rest _)) _))
                  (fn [_] (if (> (count _) 1) (rand-nth (rest _)) _))
                  (fn [_] (if (> (count _) 1) (rand-nth (rest _)) _))
                  (fn [_] (if (> (count _) 1) (rand-nth (rest _)) _))
                  (fn [_] (try (eval _) (catch Exception e _)))
                  (fn [_] `(+ ~_ 0))
                  (fn [_] `(- 0 ~_))
                  (fn [_] `(* ~_ 1))
                  (fn [_] `(Math/pow ~_ -1))
                  (fn [_] `(Math/cbrt ~_))
                  (fn [_] `(Math/exp ~_))
                  (fn [_] `(Math/log ~_))
                  (fn [_] `(Math/sqrt ~_))]
   :terminal [(fn [_] (expt 2 (/ 1 (rand))))
              (fn [_] (expt 2 (/ 1 (rand))))
              (fn [_] (rand))
              (fn [_] (rand))
              (fn [_] `(r))
              (fn [_] `(+ ~_ 0))
              (fn [_] `(* ~_ 1))]})

(defn mutate-fn [expression]
  (let [mutations (if (seq? expression)
                    (all-mutations :non-terminal)
                    (all-mutations :terminal))]
    ((rand-nth mutations) expression)))

File: src/examples/circle_area.py
import math
import random

def circle_area(radius):
    return math.pi * radius ** 2

def fitness(program):
    program_size = len(program.all_paths)
    diff_function = lambda r: circle_area(r) - program(r=r)

    return math.log(program_size + 1) * sum(diff_function(float(i)) ** 2 for i in range(30))

def try_eval(expression):
    try:
        return expression()
    except:
        return expression

non_terminal = [
    lambda e: e,
    lambda e: random.choice(e[1:]),
    lambda e: random.choice(e[1:]),
    lambda e: random.choice(e[1:]),
    lambda e: try_eval(e),
    lambda e: (float.__add__, e, 0.0),
    lambda e: (float.__sub__, 0.0, e),
    lambda e: (float.__mul__, 1.0, e),
    # lambda e: (float.__pow__, e, 1.0),
    # lambda e: (math.exp, e),
    # lambda e: (math.log, e, 1.0),
]

terminal = [
    lambda e: e,
    lambda e: random.random(),
    lambda e: random.random(),
    lambda e: random.random(),
    lambda e: 'r',
    lambda e: (float.__add__, e, 0.0),
    lambda e: (float.__mul__, 1.0, e),
]

def mutate(expression):
    mutations = non_terminal if isinstance(expression, tuple) else terminal
    return random.choice(mutations)(expression)
